src/crimson/net/Socket.cc

```diff
@@ -9,76 +9,79 @@ namespace ceph::net {
-  using consumption_result_type = typename seastar::input_stream<char>::consumption_result_type;
-
-  // consume some or all of a buffer segment
-  seastar::future<consumption_result_type> operator()(tmp_buf&& data) {
-    if (remaining >= data.size()) {
-      // consume the whole buffer
-      remaining -= data.size();
-      bl.append(buffer::create_foreign(std::move(data)));
-      if (remaining > 0) {
-        // return none to request more segments
-        return seastar::make_ready_future<consumption_result_type>(
-            seastar::continue_consuming{});
-      } else {
-        // return an empty buffer to singal that we're done
-        return seastar::make_ready_future<consumption_result_type>(
-            consumption_result_type::stop_consuming_type({}));
-      }
-    }
-    if (remaining > 0) {
-      // consume the front
-      bl.append(buffer::create_foreign(data.share(0, remaining)));
-      data.trim_front(remaining);
-      remaining = 0;
-    }
-    // give the rest back to signal that we're done
-    return seastar::make_ready_future<consumption_result_type>(
-        consumption_result_type::stop_consuming_type{std::move(data)});
-  };
-};
+inline seastar::temporary_buffer<char> sharing_split(
+  seastar::temporary_buffer<char>& buf,
+  const size_t bytes)
+{
+  auto shared_part = buf.share(0, bytes);
+  buf.trim_front(bytes);
+  return shared_part;
+}
 
 } // anonymous namespace
 
-seastar::future<bufferlist> Socket::read(size_t bytes)
+seastar::future<bufferlist> Socket::read(const size_t bytes)
 {
-  if (bytes == 0) {
-    return seastar::make_ready_future<bufferlist>();
-  }
-  r.buffer.clear();
   r.remaining = bytes;
-  return in.consume(bufferlist_consumer{r.buffer, r.remaining})
-    .then([this] {
-      if (r.remaining) { // throw on short reads
-        throw std::system_error(make_error_code(error::read_eof));
+  r.sgl.clear();
+  return seastar::do_until(
+    [this] { return r.remaining == 0; },
+    [this] {
+      if (rbuf.empty()) {
+        return in.read().then([this] (read_buffer_t&& new_rbuf) {
+          if (new_rbuf.empty()) {
+            throw std::system_error(make_error_code(error::read_eof));
+          }
+          rbuf = std::move(new_rbuf);
+          return seastar::now();
+        });
       }
-      return seastar::make_ready_future<bufferlist>(std::move(r.buffer));
-    });
+
+      const size_t round_size = std::min(r.remaining, rbuf.size());
+      r.sgl.push_back(buffer::create(sharing_split(rbuf, round_size)));
+      r.remaining -= round_size;
+      return seastar::now();
+    }
+  ).then([this] {
+    return seastar::make_ready_future<ceph::bufferlist>(std::move(r.sgl));
+  });
 }
 
-seastar::future<seastar::temporary_buffer<char>>
-Socket::read_exactly(size_t bytes) {
-  if (bytes == 0) {
-    return seastar::make_ready_future<seastar::temporary_buffer<char>>();
+seastar::future<Socket::read_buffer_t> Socket::read_exactly(size_t bytes) {
```

Comment:
`read_exactly()` doesn't really need to take the alignment parameter nor return aligned memory. I believe it's supposed to handle few cases when an interface client explicitly wants the memory to be contiguous for the sake of e.g. `reinterpret_cast<...>` to a packed `struct` (header parsing). Reading large chunks with it looks like a misuse of the interface. The key point is that `input_stream` reserves the right to decide about placement and even about size of returned chunk to particular network stack implementation.

After the changes `ProtocolV2` uses `::read()` instead of `read_exactly()` for reading msg's segments.